### PR TITLE
Handle data tiers during rolling upgrades

### PR DESCRIPTION
**Pull request by barkbay**
Fix https://github.com/elastic/cloud-on-k8s/issues/5228 by adding a predicate to control the upgrade order for the data tiers.

In the case of a circular dependency between 2 tiers the predicate does not block an upgrade, there is no guarantee for a particular order in such a situation.

### CODE REVIEW

File: pkg/apis/elasticsearch/v1/elasticsearch_config_test.go [hunk 1]
```diff
@@ -175,9 +175,61 @@ func TestConfig_HasRole(t *testing.T) {
...
+		node      *Node
+		wantRoles []NodeRole
+	}{
+		{
+			name:      "master and data",
+			node:      &Node{Roles: []string{"master", "data"}},
+			wantRoles: []NodeRole{MasterRole, DataContentRole, DataRole, DataHotRole, DataWarmRole, DataColdRole, DataFrozenRole},
+		},
+		{
+			name:      "master and data",
```
Comment (pebrc) [inline pkg/apis/elasticsearch/v1/elasticsearch_config_test.go:193]:
```suggestion
			name:      "master and data_content",
```

File: pkg/apis/elasticsearch/v1/elasticsearch_config.go [hunk 2]
```diff
@@ -8,17 +8,17 @@ import (
...
-
-	commonv1 "github.com/elastic/cloud-on-k8s/pkg/apis/common/v1"
 )
 
 type NodeRole string
 
 const (
 	DataColdRole            NodeRole = "data_cold"
 	DataContentRole         NodeRole = "data_content"
+	DataFrozenRole          NodeRole = "data_frozen"
```
Comment (pebrc) [inline pkg/apis/elasticsearch/v1/elasticsearch_config.go:21]: Not sure how we missed the frozen role and did not notice for so long. Should we also add the matching label to the pods for consistency?

File: pkg/controller/elasticsearch/driver/upgrade_predicates.go [hunk 3]
```diff
@@ -10,13 +10,98 @@ import (
...
+	case candidate.HasRole(esv1.DataWarmRole):
+		// warm tier must be upgraded after cold and frozen
+		return other.HasRole(esv1.DataColdRole) || other.HasRole(esv1.DataFrozenRole)
+	case candidate.HasRole(esv1.DataColdRole):
+		// cold tier must be upgraded after frozen
+		return other.HasRole(esv1.DataFrozenRole)
+	}
+	// frozen and content have no dependency
+	return false
+}
```
Comment (pebrc) [inline pkg/controller/elasticsearch/driver/upgrade_predicates.go:100]: I was wondering wether the dependency logic could live closer to the domain model in the `esv1` package but no strong preference.

**Review by pebrc — APPROVED** [review]
Looks good! I like how clean the dependency logic turned out despite being somewhat tricky. 👍

File: pkg/apis/elasticsearch/v1/elasticsearch_config.go (hunk 2 above)
Comment (barkbay) [reply pkg/apis/elasticsearch/v1/elasticsearch_config.go:21]: `elasticsearch.k8s.elastic.co/node-data_frozen` added in https://github.com/elastic/cloud-on-k8s/pull/5248/commits/ac81b733e712f96fc0dd6d601a4f4350ba784e2b

**Pull request merged**